src/api/java/baritone/api/pathing/goals/GoalNear.java

```diff
@@ -51,6 +54,27 @@ public double heuristic(int x, int y, int z) {
         return GoalBlock.calculate(xDiff, yDiff, zDiff);
     }
 
+    @Override
+    public double heuristic() {//TODO less hacky solution
```

Comment:
what the fuck is this

Reply:
It finds the highest heuristic that can not occur outside the goal.

Reply:
you want an ordered set then, where you can do queries based on what's greater than or less than a certain key??

Reply:
a heap, specifically??

Reply:
Since I dunno what a heap is: maybe

Reply:
I'll use a `TreeSet` because it has a better performance for add, remove and contains than `PriorityQueue` which is the only heap implementation in java (No i won't do my own for this)

Reply:
Wait, when comparing the performance for the most often used operations (add, contains, remove), `HasSet` (O(1), O(1), O(1)) is significantly better than `TreeSet` (O(log(n)), O(log(n)), O(log(n))) and `PriorityQueue` (O(log(n)), O(n), O(n)). I don't think the single access to the biggest element in the end changes much about that ordering.

Reply:
I'm not saying that you could improve on the big O complexity of a HashSet, I'm implying that the actual behavior of a Set is almost certainly not what you want here and I can't imagine how it could be. `It finds the highest heuristic that can not occur outside the goal.` You're comparing by == on these double values. That is a horrible smell and I've literally never seen a case where a hash keyed on a float/double was actually correct. "the highest heuristic that cannot occur outside the goal" you should be doing queries based on `>` instead of `==`. Because of floating point inaccuracy you can't guarantee that the heuristic calculated at dx+1 dz-1 will be equal for example. At the very minimum you need an epsilon value, and even then I wouldn't trust it.

What exactly are you trying to accomplish? Can you explain more in depth what it means to get heuristics that can't occur outside the goal??

Reply:
The full-path ETA needs the heuristic at the goal (even if that is hypothetical like for `#tunnel`) and if there are multiple possible ones I want to take the highest in order to get the lowest ETA. At the same time I need to make sure that the chosen heuristic can not occur outside the goal, so the ETA can not reach 0 before the goal is reached.
So "h can not occur outside the goal" basically means that for all combinations of x, y, z `heurisict(x,y,z) != h || isInGoal(x,y,z)` is true.
And yes, as of writing this I am thinking about whether that condition actually fits the purpose I explained above.

Reply:
And shouldn't floating point accuracy be deterministic so we can just ignore it here?

Reply:
"ETA cannot reach zero" again, you're using ≠ when you want >

A negative ETA is just as bad as a zero ETA? Why do you only care about eliminating 0 ETA and not - ETA?

"h cannot occur outside the goal" should actually be "h or lower cannot occur outside the goal".

`heurisict(x,y,z) != h || isInGoal(x,y,z)` should instead be `heurisict(x,y,z) > h || isInGoal(x,y,z)`

Reply:
That's what the last sentence was about. The current solution allows for e.g. 5 being returned although e.g. 3 is a possible heuristic outside the goal. In addition to that it fails when there is no in-goal-only heuristic.
I guess I will change this to start with `h = Double.MAX` and then do `if (!isInGoal(x, y, z)) h = Math.min(h, heuristic(x, y, z)` for each position inside the box. That way it always has a value that is guaranteed to be lower than all not-in-goal heuristics.

Reply:
We could also see it as "The heuristic says you could already be there. Why are you wasting time?" because when the heuristic is x at a certain point inside the we should not need to get closer to the goal than x.
I'll still change it because the new solution looks nicer and has better performance which is actually an issue for `#follow` with high followRange (100 or so, not really likely to happen).